File: 07/claude-opus-4.5/typescript/solution.ts
import * as fs from 'fs';

const input = fs.readFileSync('input.txt', 'utf-8').trim();
const lines = input.split('\n');
const rows = lines.length;
const cols = rows > 0 ? lines[0].length : 0;

// Find starting position S
let startRow = 0, startCol = 0;
for (let r = 0; r < rows; r++) {
    const c = lines[r].indexOf('S');
    if (c !== -1) {
        startRow = r;
        startCol = c;
        break;
    }
}

// Part 1: Count total splits
let splitCount = 0;
let currentBeams = new Set<number>([startCol]);

for (let row = startRow + 1; row < rows && currentBeams.size > 0; row++) {
    const nextBeams = new Set<number>();
    for (const col of currentBeams) {
        if (col < 0 || col >= cols) continue;
        const cell = lines[row][col];
        if (cell === '^') {
            splitCount++;
            if (col - 1 >= 0) nextBeams.add(col - 1);
            if (col + 1 < cols) nextBeams.add(col + 1);
        } else {
            nextBeams.add(col);
        }
    }
    currentBeams = nextBeams;
}

console.log(`Day 07 Part 1: ${splitCount}`);

// Part 2: Count timelines (each particle takes both paths at each splitter)
let particleCounts = new Map<number, bigint>([[startCol, 1n]]);

for (let row = startRow + 1; row < rows && particleCounts.size > 0; row++) {
    const nextCounts = new Map<number, bigint>();
    for (const [col, count] of particleCounts) {
        if (col < 0 || col >= cols) continue;
        const cell = lines[row][col];
        if (cell === '^') {
            if (col - 1 >= 0) {
                nextCounts.set(col - 1, (nextCounts.get(col - 1) || 0n) + count);
            }
            if (col + 1 < cols) {
                nextCounts.set(col + 1, (nextCounts.get(col + 1) || 0n) + count);
            }
        } else {
            nextCounts.set(col, (nextCounts.get(col) || 0n) + count);
        }
    }
    particleCounts = nextCounts;
}

let part2 = 0n;
for (const count of particleCounts.values()) {
    part2 += count;
}
console.log(`Day 07 Part 2: ${part2}`);

File: 07/claude-opus-4.5/python/solution.py
with open("input.txt") as f:
    lines = f.read().strip().split('\n')

grid = [list(line) for line in lines]
rows = len(grid)
cols = len(grid[0]) if rows > 0 else 0

# Find starting position S
start_row, start_col = 0, 0
for r in range(rows):
    for c in range(cols):
        if grid[r][c] == 'S':
            start_row, start_col = r, c
            break

# Part 1: Count total splits
split_count = 0
current_beams = {start_col}  # Set of columns with active beams

for row in range(start_row + 1, rows):
    if not current_beams:
        break
    next_beams = set()
    for col in current_beams:
        if col < 0 or col >= cols:
            continue
        cell = grid[row][col]
        if cell == '^':
            # Beam hits a splitter - count the split
            split_count += 1
            # Create new beams to the left and right
            if col - 1 >= 0:
                next_beams.add(col - 1)
            if col + 1 < cols:
                next_beams.add(col + 1)
        else:
            # Beam continues downward
            next_beams.add(col)
    current_beams = next_beams

print(f"Day 07 Part 1: {split_count}")

# Part 2: Count timelines (each particle takes both paths at each splitter)
particle_counts = {start_col: 1}

for row in range(start_row + 1, rows):
    if not particle_counts:
        break
    next_counts = {}
    for col, count in particle_counts.items():
        if col < 0 or col >= cols:
            continue
        cell = grid[row][col]
        if cell == '^':
            # Each particle splits into two
            if col - 1 >= 0:
                next_counts[col - 1] = next_counts.get(col - 1, 0) + count
            if col + 1 < cols:
                next_counts[col + 1] = next_counts.get(col + 1, 0) + count
        else:
            next_counts[col] = next_counts.get(col, 0) + count
    particle_counts = next_counts

part2 = sum(particle_counts.values())
print(f"Day 07 Part 2: {part2}")
